include/pybind11/detail/eigen_tensor.h

```diff
@@ -0,0 +1,383 @@
+        }
+        return cast_impl(&src, policy, parent);
+    }
+
+    static handle cast(const Type &src, return_value_policy policy, handle parent) {
+        if (policy == return_value_policy::automatic
+            || policy == return_value_policy::automatic_reference) {
+            policy = return_value_policy::copy;
+        }
+        return cast(&src, policy, parent);
+    }
+
+    static handle cast(Type *src, return_value_policy policy, handle parent) {
+        if (policy == return_value_policy::automatic) {
+            policy = return_value_policy::take_ownership;
+        } else if (policy == return_value_policy::automatic_reference) {
+            policy = return_value_policy::reference;
+        }
+        return cast_impl(src, policy, parent);
+    }
+
+    static handle cast(const Type *src, return_value_policy policy, handle parent) {
+        if (policy == return_value_policy::automatic) {
+            policy = return_value_policy::take_ownership;
+        } else if (policy == return_value_policy::automatic_reference) {
+            policy = return_value_policy::reference;
+        }
+        return cast_impl(src, policy, parent);
+    }
+
+    template <typename C>
+    static handle cast_impl(C *src, return_value_policy policy, handle parent) {
+        object parent_object;
+        bool writeable = false;
+        switch (policy) {
+            case return_value_policy::move:
+                if (std::is_const<C>::value) {
+                    pybind11_fail("Cannot move from a constant reference");
+                }
+                {
+                    Eigen::aligned_allocator<Type> allocator;
+                    Type *copy = ::new (allocator.allocate(1)) Type(std::move(*src));
+                    src = copy;
+                }
+
+                parent_object = capsule(src, [](void *ptr) {
+                    Eigen::aligned_allocator<Type> allocator;
+                    Type *copy = (Type *) ptr;
+                    copy->~Type();
+                    allocator.deallocate(copy, 1);
+                });
+                writeable = true;
+                break;
+
+            case return_value_policy::take_ownership:
+                if (std::is_const<C>::value) {
+                    pybind11_fail("Cannot take ownership of a const reference");
+                }
+                parent_object = capsule(src, [](void *ptr) { delete (Type *) ptr; });
+                writeable = true;
+                break;
+
+            case return_value_policy::copy:
+                parent_object = {};
+                writeable = true;
+                break;
+
+            case return_value_policy::reference:
+                parent_object = none();
+                writeable = !std::is_const<C>::value;
+                break;
+
+            case return_value_policy::reference_internal:
+                // Default should do the right thing
+                parent_object = reinterpret_borrow<object>(parent);
```

Comment:
You might want to cast to object instead of reinterpret borrow. I don't think reinterpert_borrow will error if the handle is null. (one difference between handle and object is a nullptr is not a valid for the pointer of an object (at least when it is initialized)).